README.markdown
# OpenLink ODBC Data Adapter for Ruby on Rails

(C) 2008 OpenLink Software


# Status

**23-Nov-2015**
  * Update fetch query results

**19-Nov-2015**

Update to preserve limits on queries
Update to add query comments

**26-Jun-2015**

Added support for Rails 4.2

**12-Sep-2012**

Added support for Rails / ActiveRecord 3.1.x and 3.2

**22-Nov-2008**

Added new install information

**23-Apr-2008**

The adapter has been updated to support Rails 2.0.2 / ActiveRecord 2.0.2
The new adapter (v2.0) is not recommended for use with earlier releases of
Rails or ActiveRecord.

* For Rails 2.0.2 / ActiveRecord 2.0.2 or later, use odbc-rails v2.0
* For Rails 1.2.x / ActiveRecord 1.15.x, use odbc-rails v1.5
* For Rails 1.1.x / ActiveRecord 1.14.x, use odbc-rails v1.3

The configuration instructions for v1.4 or earlier differ slightly from those
included here for v2.0. Please refer to the documentation packaged with the
respective adapter when using the previous releases.

Added support for DSN-less connections (thanks to Ralf Vitasek).
Added support for SQLAnywhere (thanks to Bryan Lahartinger).

**23-Feb-2007**

The adapter has been updated to support Rails 1.2.x / ActiveRecord 1.15.x.
The new adapter (v1.4) is not recommended for use with Rails 1.1 / ActiveRecord
1.14.x. Please use v1.3 of the adapter with Rails 1.1.

The driver now supports the new ActiveRecord :decimal type and an
:emulate_booleans connection option. See http://odbc-rails.rubyforge.org for
more information about this option.

**09-Jan-2007**

The adapter accompanying this note is a Generic ODBC Adapter for Ruby on Rails
being developed by **OpenLink&nbsp;Software**[http://www.openlinksw.com].

The aim of this development is to provide a single ODBC-based adapter
capable of supporting the most popular DBMSes, in contrast to the
current approach in the Rails community of each database requiring
its own adapter.

It currently supports Ingres r3, Informix 9.3 or later, Oracle 10g,
MySQL 5 and OpenLink's Virtuoso
(Open&nbsp;Source&nbsp;Edition[http://virtuoso.openlinksw.com]),
SQL Server 2000/2005, Sybase ASE 15, DB2 v9, Progress v8/9/10 and PostgreSQL 8.2.

Testing to date has been limited to the ROR 'Expenses' sample
application described at http://developer.apple.com/tools/rubyonrails.html
and the ActiveRecord test modules base_test.rb and migration_test.rb.

Testing has been performed on Linux, Windows and Mac OS X using
OpenLink's own ODBC drivers and the native Virtuoso ODBC client.


# Pre-Requisites

The OpenLink ODBC Adapter for Ruby on Rails needs the
ActiveRecord[http://wiki.rubyonrails.com/rails/pages/ActiveRecord]
package installed.

The Adapter also requires Christian Werner's
Ruby&nbsp;ODBC&nbsp;Bridge[http://www.ch-werner.de/rubyodbc]
(release 0.9991 or later), to bridge to an underlying ODBC driver.


# Contents

In the accompanying sources, the lib directory structure is equivalent
to the lib directory located under ACTIVE_RECORD_ROOT in your main
Ruby tree.

On Windows ACTIVE_RECORD_ROOT will be something like:

    c:\ruby\lib\ruby\gems\1.8\gems\activerecord-x.y.z

On Unix and Mac OS X ACTIVE_RECORD_ROOT will be something like:

    /usr/lib/ruby/gems/1.8/gems/activerecord-x.y.z
or
    /usr/local/lib/ruby/gems/1.8/gems/activerecord-x.y.z

On Mac OS X using Locomotive ACTIVE_RECORD_ROOT will be
something like:

    /Applications/Locomotive2/bundles/rails112.locobundle/i386/lib/ruby/gems/1.8/gems/activerecord-x.y.z

lib/ contains the files which constitute the new ODBC adapter.

test/ contains the fixture definitions for testing the adapter.

support/odbc_rails.diff contains some patches to ActiveRecord tests.

The patches to base_test.rb and migration_test.rb are not essential.
However, they modify or bypass certain tests to cope with limitations
of particular databases. Other developers have previously modified
the tests similarly to cope with limitations of other Rails-supported
databases. The patched versions of files touched by odbc_rails.diff can
be found in support/test.

# Installation

There are 3 ways to install the ODBC Adapter package: either as a gem
(recommended), a plugin, or by running the custom installation script. Pick
one of the following, depending on whether you want the adapter to be
available system-wide or just within a particular Rails project.

# Installation as a Gem (recommended)

* If you haven't done so already first add github to rubygems with:

  gem sources -a http://gems.github.com

* Install the odbc-rails gem by running:

  gem install -r dosire-activerecord-odbc-adapter --include-dependencies

# Installation as a Plugin

Installing the OpenLink ODBC Data Adapter as a plugin makes it available to
a particular Rails application only.

# Automatic Plugin installation

The adapter can be automatically installed as a plugin by navigating to the
root of your Rails application and typing either:

    script/plugin install odbc-rails
or
    script/plugin install http://odbc-rails.rubyforge.org/plugins/odbc-rails

On Windows, replace <tt>script/plugin</tt> by <tt>ruby script/plugin</tt>.
e.g. <tt>ruby script/plugin install odbc-rails</tt>


# Manual Plugin Installation

You can also install the plugin manually by unpacking the sources into the
vendor/plugins directory of your Rails application.

# Custom installation script

When using <tt>rake install</tt> or running the install_odbc.rb script by
hand, the script tries to determine the location of the activerecord
package within your Ruby installation and allows you to overrule the
directory it has found with a directory of your choice. If more than one
version of activerecord is found, the default will be the latest version.

On Mac OS X the installation script is also capable of locating Locomotive,
if installed, and will ask to install in there.

The install_odbc.rb script also verifies that the pre-requisites are
installed.

If all else fails this document also describes the steps to install the
ODBC adapter into activerecord manually.


# Using rake

The simplest way to install the adapter is by using the following
command as root:

    # rake install

or if your system supports sudo:

    $ sudo rake install


The Rakefile also defines some targets specifically for developers:

    $ rake rdoc         # Generate rdoc documentation in ~/doc

    $ rake package	# Create distribution package in ~/distrib

    $ rake clean	# Remove generated files and directories


# Running installation script

The second way of installing the adapter is running the
<tt>install_odbc.rb</tt> script as root.

    # ruby install_odbc.rb

or if your system supports sudo:

    $ sudo ruby install_odbc.rb


# Installing files manually

The third way of installing the adapter is by performing the installation
steps yourself.

* Copy odbc_adapter.rb to the ACTIVE_RECORD_ROOT/lib/active_record/connection_adapters/ directory

* Copy the odbcext_*.rb files to the ACTIVE_RECORD_ROOT/lib/active_record/vendor/ directory

# Configuring the Adapter

Examples of the required connection parameters can be found in

   test/connections/native_odbc/connection.rb

If you enable call-tracing by setting :trace #> true,  specify the
logger output file as illustrated in connection.rb, e.g.

    ActiveRecord::Base.logger # Logger.new("debug_odbc.log")

# License

The OpenLink ODBC Adapter for Ruby on Rails is released under the
MIT license as detailed in the file COPYING.


# Submitting fixes and enhancements

Patches and new contributions can be submitted as diffs from the
current CVS archive by:

    $ cvs add newfiles

    $ cvs -z3 diff -uN > diffs

Patches and  contributions can be send to the OpenLink iODBC source
archive mainainter at mailto:iodbc@openlinksw.com to be included
in the next distribution. Please provide accompanying documentation
on which bugs are fixed or new features are introduced.


NEWS
Nov 23, 2015, V4.2.2
  * Update fetch query results

Nov 19, 2015, V4.2.1
  * Update to preserve limits on queries
  * Update to add query comments

Jun 26, 2015, V4.2.0
  * Updated to support Rails 4.2

September 12, 2012, V2.1
  * Changes for Rails / ActiveRecord 3.1 and 3.2

April 23, 2008, V2.0:
  * Changes for Rails 2.0.x / ActiveRecord 2.0.x
  * Added support for DSN-less connections
  * Added support for SQLAnywhere

February 27, 2007, V1.4:
  * Changes for Rails 1.2.x / ActiveRecord 1.15.x
  * Added support for :decimal type
  * Added :emulate_booleans connection option

January 9, 2007, V1.3:
  * Added support for PostgreSQL

December 11, 2006, V1.2:
  * Added option to install adapter as a gem
  * Added option to install adapter as a plugin
  * Added support for Progress 8 (SQL 89)
  * Added support for Progress 9 and 10 (SQL 92)

October 18, 2006, V1.1:
  * Added support for DB2, MySQL, Sybase and SQL Server
  * Fixed initial unicode support

August 23 2006, V1.0:
  * Initial archive


support/rake/rails_plugin_package_task.rb
#!/usr/bin/env ruby

# Copyright (c) 2006 by Zak Mandhro
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of this software
# and associated documentation files (the "Software"), to deal in the Software without restriction,
# including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all copies or substantial
# portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
# LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
# EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
# USE OR OTHER DEALINGS IN THE SOFTWARE.
#
# = Rails Plug-in Package Task
#
# RailsPluginPackageTask is a rake task designed to automate the publishing of
# Ruby on Rails plug-ins. The Rails plug-in installer _RecursiveHTTPFetcher_ makes
# certain assumptions about the web servers that does not hold through from
# server to server, for example:
#
# * Server generates an index page with links
# * All links to plug-in files are relative links
# * Folder links end with a forward slash (used to recurse)
#
# RubyForge web server is an example of where these assupmtions don't hold
# true. As a result, you can not simply copy your files to a web server
# and expect Rails HTTP plugin installer to just work.
#
# This Rake task helps fill the gap by complying to the plug-in scripts assumptions.
# Following the Rake package task conventions, it defines the "rails_plugin" task
# that recurses through your _package_files_, generates compliant index.html for
# each folder (that contains a file), and creates a directory structure that you
# can publish as a set for your plugin.
#
# = Example
#
# The following example uses the Rake::RailsPluginPackageTask to create
# the package. It then uses the Rake::SshDirPublisher to publish the plugin
# directory to RubyForge.
# #
#    Rake::RailsPluginPackageTask.new(ProjectInfo[:name], ProjectInfo[:version]) do |p|
#      p.package_files = PluginPackageFiles
#      p.plugin_files = FileList["rails_plugin/**/*"]
#      p.extra_links = {"Project page"=>ProjectInfo[:homepage],
#        "Author: #{ProjectInfo[:author_name]}"=>ProjectInfo[:author_link]}
#      p.verbose = true
#    end
#    task :rails_plugin=>:clobber
#
#    desc "Publish Ruby on Rails plug-in on RubyForge"
#    task :release_plugin=>:rails_plugin do |task|
#      pub = Rake::SshDirPublisher.new("#{RubyForgeConfig[:user_name]}@rubyforge.org",
#    	"/var/www/gforge-projects/#{RubyForgeConfig[:unix_name]}",
#    	"pkg/rails_plugin")
#      pub.upload()
#    end


require 'rake'
require 'rake/tasklib'

module Rake
  # RailsPluginPackageTask defines the "rails_plugin" task
  # that recurses through your _package_files_, generates compliant index.html for
  # each folder (that contains a file), and creates a directory structure that you
  # can publish as a set for your plugin.
  #
  # Noteworthy attributes:
  #
  # [package_dir]  Directory to store the package. Default 'pkg/rails_plugin'
  #
  # [package_dir]  Files to include in the plugin.
  #
  # [extra_links]  Links to put on every generated index page. Can be a hash, e.g.
  #                {"Home"=>"http://roxml.rubyforge.org"}, an array of strings or
  #                a single string.
  #
  # [plugin_files]  Files to be placed in the root folder of the plug-in, e.g.
  #                 init.rb. All files that are in the root of _package_dir_
  #                 will also be placed in the root of the plug-in.
  #
  class RailsPluginPackageTask < TaskLib
    # Name of plug-in or application
    attr_accessor :name
    # Version of plugin - distribution folder will be name_version
    attr_accessor :version
    # Directory used to store the package files (default is 'pkg/rails_plugin').
    attr_accessor :package_dir
    # Files to be stored in the package
    attr_accessor :package_files
    # Files to go into the root of the plug-in folder (e.g. init.rb)
    attr_accessor :plugin_files
    # Homepage for more information
    attr_accessor :extra_links
    # Verbose [true | false]
    attr_accessor :verbose

    # Create the "rails_plugin" task
    def initialize(name=nil, version=nil)
      init(name, version)
      yield self if block_given?
      define unless name.nil?
    end

    # Initialize with defaults
    def init(name, version)
      @name = name
      @version = version
      @extra_links = nil
      @package_files = Rake::FileList.new
      @plugin_files = Rake::FileList.new
      @package_dir = 'pkg/rails_plugin'
      @folders = {}
      @verbose = false
    end

    # Define the rails_plugin task
    def define
      desc "Create Rails plug-in package"
      task :rails_plugin do
        @dest = "#@package_dir/#{@name}_#{@version}"
        makedirs(@dest,:verbose=>false)
        @plugin_files.each do |fn|
          cp(fn, @dest,:verbose=>false)
          add_file(File.basename(fn))
        end

        @package_files.each do |fn|
          puts ". #{fn}" if verbose
          f = File.join(@dest, fn)
          fdir = File.dirname(f)
          unless File.exist?(fdir)
            mkdir_p(fdir,:verbose=>false)
            add_folder("#{fdir}/")
          end
          if File.directory?(fn)
            mkdir_p(f,:verbose=>false)
            add_folder("#{fn}/")
          else
            cp(fn, f, :verbose=>false)
            add_file(fn)
          end
        end

        generate_index_files()
      end
    end

    # Generate the index.html files
    def generate_index_files
      @folders.each do |folder, files|
        puts " + Creating #{@dest}/#{folder}/index.html" if @verbose
        File.open("#{@dest}/#{folder}/index.html", "w") do |index|
          title = "Rails Plug-in for #@name #@version"
          index.write("<html><head><title>#{title}</title></head>\n")
          index.write("<body>\n")
          index.write("<h2>#{title}</h2>\n")
          extra_links = create_extra_links()
          index.write("<p>#{extra_links}</p>\n") if extra_links
          files.each { |fn|
            puts("  - Adding #{fn}") if @verbose
            index.write("&nbsp;&nbsp;<a href=\"#{fn}\">#{fn}</a><br/>\n")
          }
          index.write("<hr size=\"1\"/><p style=\"font-size: x-small\">Generated with RailsPluginPackageTask<p>")
          index.write("</body>\n")
          index.write("</html>\n")
        end
      end
    end

  private
    # Add a file to the folders hash
    def add_file(filename)
      dir = File.dirname(filename).gsub("#{@dest}",".")
      fn = File.basename(filename)
      folder = @folders[dir] || @folders[dir]=[]
      folder << fn
    end

    # Add a folder to the folders hash
    def add_folder(folder_name)
      dir = File.dirname(folder_name).gsub("#{@dest}",".").gsub("./","")
      fn = File.basename(folder_name) + "/"
      folder = @folders[dir] || @folders[dir]=[]
      folder << fn
    end

    # Create the anchor tag for extra links
    def create_extra_links
      return nil unless @extra_links
      x_links = ""
      if (@extra_links.class==Hash)
        @extra_links.each do |k,v|
          x_links << "<a href=\"#{v}\">#{k}</a>&nbsp;"
        end
      elsif (@extra_links.class==Array)
        @extra_links.each do |link|
          x_links << "<a href=\"#{link}\">#{link}</a>&nbsp;"
        end
      else
        x_links = "<a href=\"#{@extra_links.to_s}\">#{@extra_links.to_s}</a>"
      end
      return x_links
    end
  end
end


support/lib/active_record/connection_adapters/abstract/schema_definitions.rb
require 'parsedate'

module ActiveRecord
  module ConnectionAdapters #:nodoc:
    # An abstract definition of a column in a table.
    class Column
      attr_reader :name, :default, :type, :limit, :null, :sql_type
      attr_accessor :primary

      # Instantiates a new column in the table.
      #
      # +name+ is the column's name, as in <tt><b>supplier_id</b> int(11)</tt>.
      # +default+ is the type-casted default value, such as <tt>sales_stage varchar(20) default <b>'new'</b></tt>.
      # +sql_type+ is only used to extract the column's length, if necessary.  For example, <tt>company_name varchar(<b>60</b>)</tt>.
      # +null+ determines if this column allows +NULL+ values.
      def initialize(name, default, sql_type = nil, null = true)
        @name, @type, @null = name, simplified_type(sql_type), null
        @sql_type = sql_type
        # have to do this one separately because type_cast depends on #type
        @default = type_cast(default)
        @limit   = extract_limit(sql_type) unless sql_type.nil?
        @primary = nil
        @text    = [:string, :text].include? @type
        @number  = [:float, :integer].include? @type
      end

      def text?
        @text
      end

      def number?
        @number
      end

      # Returns the Ruby class that corresponds to the abstract data type.
      def klass
        case type
          when :integer       then Fixnum
          when :float         then Float
          when :datetime      then Time
          when :date          then Date
          when :timestamp     then Time
          when :time          then Time
          when :text, :string then String
          when :binary        then String
          when :boolean       then Object
        end
      end

      # Casts value (which is a String) to an appropriate instance.
      def type_cast(value)
        return nil if value.nil?
        case type
          when :string    then value
          when :text      then value
          when :integer   then value.to_i rescue value ? 1 : 0
          when :float     then value.to_f
          when :datetime  then self.class.string_to_time(value)
          when :timestamp then self.class.string_to_time(value)
          when :time      then self.class.string_to_dummy_time(value)
          when :date      then self.class.string_to_date(value)
          when :binary    then self.class.binary_to_string(value)
          when :boolean   then self.class.value_to_boolean(value)
          else value
        end
      end

      def type_cast_code(var_name)
        case type
          when :string    then nil
          when :text      then nil
          when :integer   then "(#{var_name}.to_i rescue #{var_name} ? 1 : 0)"
          when :float     then "#{var_name}.to_f"
          when :datetime  then "#{self.class.name}.string_to_time(#{var_name})"
          when :timestamp then "#{self.class.name}.string_to_time(#{var_name})"
          when :time      then "#{self.class.name}.string_to_dummy_time(#{var_name})"
          when :date      then "#{self.class.name}.string_to_date(#{var_name})"
          when :binary    then "#{self.class.name}.binary_to_string(#{var_name})"
          when :boolean   then "#{self.class.name}.value_to_boolean(#{var_name})"
          else nil
        end
      end

      # Returns the human name of the column name.
      #
      # ===== Examples
      #  Column.new('sales_stage', ...).human_name #=> 'Sales stage'
      def human_name
        Base.human_attribute_name(@name)
      end

      # Used to convert from Strings to BLOBs
      def self.string_to_binary(value)
        value
      end

      # Used to convert from BLOBs to Strings
      def self.binary_to_string(value)
        value
      end

      def self.string_to_date(string)
        return string unless string.is_a?(String)
        date_array = ParseDate.parsedate(string)
        # treat 0000-00-00 as nil
        Date.new(date_array[0], date_array[1], date_array[2]) rescue nil
      end

      def self.string_to_time(string)
        return string unless string.is_a?(String)
        time_array = ParseDate.parsedate(string)[0..5]
        # treat 0000-00-00 00:00:00 as nil
        Time.send(Base.default_timezone, *time_array) rescue nil
      end

      def self.string_to_dummy_time(string)
        return string unless string.is_a?(String)
        time_array = ParseDate.parsedate(string)
        # pad the resulting array with dummy date information
        time_array[0] = 2000; time_array[1] = 1; time_array[2] = 1;
        Time.send(Base.default_timezone, *time_array) rescue nil
      end

      # convert something to a boolean
      def self.value_to_boolean(value)
        return value if value==true || value==false
        case value.to_s.downcase
        when "true", "t", "1" then true
        else false
        end
      end

    private
        def extract_limit(sql_type)
          $1.to_i if sql_type =~ /\((.*)\)/
        end

        def simplified_type(field_type)
          case field_type
            when /int/i
              :integer
            when /float|double|decimal|numeric/i
              :float
            when /datetime/i
              :datetime
            when /timestamp/i
              :timestamp
            when /time/i
              :time
            when /date/i
              :date
            when /clob/i, /text/i
              :text
            when /blob/i, /binary/i
              :binary
            when /char/i, /string/i
              :string
            when /boolean/i
              :boolean
          end
        end
    end

    class IndexDefinition < Struct.new(:table, :name, :unique, :columns) #:nodoc:
    end

    class ColumnDefinition < Struct.new(:base, :name, :type, :limit, :default, :null) #:nodoc:
      def to_sql
        column_sql = "#{base.quote_column_name(name)} #{type_to_sql(type.to_sym, limit)}"
        add_column_options!(column_sql, :null => null, :default => default)
        column_sql
      end
      alias to_s :to_sql

      private
        def type_to_sql(name, limit)
          base.type_to_sql(name, limit) rescue name
        end

        def add_column_options!(sql, options)
          base.add_column_options!(sql, options.merge(:column => self))
        end
    end

    # Represents a SQL table in an abstract way.
    # Columns are stored as ColumnDefinition in the #columns attribute.
    class TableDefinition
      attr_accessor :columns

      def initialize(base)
        @columns = []
        @base = base
      end

      # Appends a primary key definition to the table definition.
      # Can be called multiple times, but this is probably not a good idea.
      def primary_key(name)
        column(name, :primary_key)
      end

      # Returns a ColumnDefinition for the column with name +name+.
      def [](name)
        @columns.find {|column| column.name.to_s == name.to_s}
      end

      # Instantiates a new column for the table.
      # The +type+ parameter must be one of the following values:
      # <tt>:primary_key</tt>, <tt>:string</tt>, <tt>:text</tt>,
      # <tt>:integer</tt>, <tt>:float</tt>, <tt>:datetime</tt>,
      # <tt>:timestamp</tt>, <tt>:time</tt>, <tt>:date</tt>,
      # <tt>:binary</tt>, <tt>:boolean</tt>.
      #
      # Available options are (none of these exists by default):
      # * <tt>:limit</tt>:
      #   Requests a maximum column length (<tt>:string</tt>, <tt>:text</tt>,
      #   <tt>:binary</tt> or <tt>:integer</tt> columns only)
      # * <tt>:default</tt>:
      #   The column's default value.  You cannot explicitely set the default
      #   value to +NULL+.  Simply leave off this option if you want a +NULL+
      #   default value.
      # * <tt>:null</tt>:
      #   Allows or disallows +NULL+ values in the column.  This option could
      #   have been named <tt>:null_allowed</tt>.
      #
      # This method returns <tt>self</tt>.
      #
      # ===== Examples
      #  # Assuming def is an instance of TableDefinition
      #  def.column(:granted, :boolean)
      #    #=> granted BOOLEAN
      #
      #  def.column(:picture, :binary, :limit => 2.megabytes)
      #    #=> picture BLOB(2097152)
      #
      #  def.column(:sales_stage, :string, :limit => 20, :default => 'new', :null => false)
      #    #=> sales_stage VARCHAR(20) DEFAULT 'new' NOT NULL
      def column(name, type, options = {})
        column = self[name] || ColumnDefinition.new(@base, name, type)
        column.limit = options[:limit] || native[type.to_sym][:limit] if options[:limit] or native[type.to_sym]
        column.default = options[:default]
        column.null = options[:null]
        @columns << column unless @columns.include? column
        self
      end

      # Returns a String whose contents are the column definitions
      # concatenated together.  This string can then be pre and appended to
      # to generate the final SQL to create the table.
      def to_sql
        @columns * ', '
      end

      private
        def native
          @base.native_database_types
        end
    end
  end
end


support/pack_odbc.rb
#
#  $Id: pack_odbc.rb,v 1.2 2006/08/21 23:37:17 source Exp $
#
#  OpenLink ODBC Adapter for Ruby on Rails
#  Copyright (C) 2006 OpenLink Software
#
#  Permission is hereby granted, free of charge, to any person obtaining
#  a copy of this software and associated documentation files (the
#  "Software"), to deal in the Software without restriction, including
#  without limitation the rights to use, copy, modify, merge, publish,
#  distribute, sublicense, and/or sell copies of the Software, and to
#  permit persons to whom the Software is furnished to do so, subject
#  to the following conditions:
#
#  The above copyright notice and this permission notice shall be
#  included in all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
#  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
#  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
#  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
#  ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
#  CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
#  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

#
# Copies Rails ODBC adapter files from development ActiveRecord tree to a
# staging area ready for packaging.
#

require 'fileutils'
require 'find'
require 'ftools'
require 'rdoc/rdoc'

# Development ActiveRecord source tree
$AR_DEV_ROOT = "/data/radrails/carl/activerecord_svn_co"
# Staging area
$PACK_ROOT = "/tmp/rails_odbc_pack"
# Location of miscellaneous files not in $AR_DEV_ROOT
$MISC_ROOT = "/dev/rails_odbc"

miscFiles = [
 "install_odbc.rb",
 "pack_odbc.rb",
 "readme.html"
]

# Stems of new .sql files added to $AR_DEV_ROOT/test/fixtures/db_definitions
# !! UPDATE AS SUPPORT FOR OTHER DATABASES IS ADDED TO ODBC ADAPTER
wantedBasenamePrefixes = [
 "informix", "ingres", "virtuoso", "oracle_odbc"
]

# Files in public ActiveRecord source tree which have been modified for
# ODBC adapter
modifiedFiles = [
  "./test/base_test.rb",
  "./test/migration_test.rb",
  "./lib/active_record/connection_adapters/abstract/schema_definitions.rb"
]

raise Exception, "Directory doesn't exist: #{$AR_DEV_ROOT}" if !File.exist?($AR_DEV_ROOT)

# Create directory tree under $PACK_ROOT
FileUtils.mkdir_p($PACK_ROOT)
FileUtils.mkdir_p(File.join($PACK_ROOT, 'new_files/lib/active_record/connection_adapters'))
FileUtils.mkdir_p(File.join($PACK_ROOT, 'new_files/lib/active_record/vendor'))
FileUtils.mkdir_p(File.join($PACK_ROOT, 'new_files/test/fixtures/db_definitions'))
FileUtils.mkdir_p(File.join($PACK_ROOT, 'new_files/test/connections/native_odbc'))
FileUtils.mkdir_p(File.join($PACK_ROOT,
    'modified_files/lib/active_record/connection_adapters/abstract'))
FileUtils.mkdir_p(File.join($PACK_ROOT, 'modified_files/test'))

# Generate RDoc's
Dir.glob($AR_DEV_ROOT + "/**/odbc_adapter.rb") { |f|
  FileUtils.cp(f, $PACK_ROOT)
  Dir.chdir($PACK_ROOT)
  FileUtils.rmtree('doc')
  r = RDoc::RDoc.new
  r.document(['-q', 'odbc_adapter.rb'])
  FileUtils.rm('odbc_adapter.rb')
}

# Copy new files into $PACK_ROOT
# i.e. files not in the current ActiveRecord distribution
Dir.chdir($AR_DEV_ROOT)
Find.find(".") { |f|
  if File.fnmatch("./**/odbc*.rb" , f)
    FileUtils.cp(f, File.join($PACK_ROOT, "new_files", *f.split(/\//)),
    :verbose => true)
  end
}

f = "./test/connections/native_odbc/connection.rb"
FileUtils.cp(f, File.join($PACK_ROOT, "new_files", *f.split(/\//)),
    :verbose => true)

Find.find(".") { |f|
  if File.fnmatch("./**/*.sql" , f)
    basename = File.basename(f, ".sql")
    wantedBasenamePrefixes.each do |prefix|
      if basename.match("^#{prefix}")
        FileUtils.cp(f, File.join($PACK_ROOT, "new_files", *f.split(/\//)),
            :verbose => true)
      end
    end
  end
}

modifiedFiles.each do |f|
  FileUtils.cp(f, File.join($PACK_ROOT, "modified_files", *f.split(/\//)),
      :verbose => true)
end

miscFiles.each do |f|
  FileUtils.cp(File.join($MISC_ROOT, f), $PACK_ROOT, :verbose => true)
end
